File: server/sessionRouter.ts
 import { WebSocket } from 'ws'

export type ParticipantRole = 'customer' | 'consultant'

export interface SignalMessage {
  event: 'join' | 'leave' | 'offer' | 'answer' | 'ice-candidate' | 'hangup'
  sessionId: string
  timestamp: number
  role?: ParticipantRole
  source?: string
  reason?: string
  sdp?: unknown
  candidate?: unknown
}

interface ParticipantState {
  sessionId: string | null
  role: ParticipantRole | null
}

export class SessionRouter {
  private readonly sessions = new Map<string, Set<WebSocket>>()
  private readonly participants = new WeakMap<WebSocket, ParticipantState>()

  register(socket: WebSocket): void {
    this.participants.set(socket, { sessionId: null, role: null })
  }

  remove(socket: WebSocket): void {
    const state = this.participants.get(socket)
    this.participants.delete(socket)

    if (!state?.sessionId) {
      return
    }

    const peers = this.sessions.get(state.sessionId)
    if (!peers) {
      return
    }

    peers.delete(socket)
    if (peers.size === 0) {
      this.sessions.delete(state.sessionId)
    }
  }

  join(socket: WebSocket, sessionId: string, role: ParticipantRole): void {
    this.remove(socket)

    const peers = this.sessions.get(sessionId) ?? new Set<WebSocket>()
    peers.add(socket)
    this.sessions.set(sessionId, peers)

    this.participants.set(socket, { sessionId, role })
  }

  leave(socket: WebSocket): void {
    this.remove(socket)
    this.participants.set(socket, { sessionId: null, role: null })
  }

  broadcastToPeers(sender: WebSocket, message: SignalMessage): number {
    const state = this.participants.get(sender)
    if (!state?.sessionId) {
      return 0
    }

    const peers = this.sessions.get(state.sessionId)
    if (!peers) {
      return 0
    }

    const payload = JSON.stringify(message)
    let delivered = 0

    for (const peer of peers) {
      if (peer === sender || peer.readyState !== WebSocket.OPEN) {
        continue
      }

      peer.send(payload)
      delivered += 1
    }

    return delivered
  }
}

File: backend/src/main/java/com/videoshop/backend/signaling/SessionRouter.java
package com.videoshop.backend.signaling;

import org.springframework.stereotype.Component;
import org.springframework.web.socket.TextMessage;
import org.springframework.web.socket.WebSocketSession;

import java.io.IOException;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

@Component
public class SessionRouter {

	private final Map<String, Set<WebSocketSession>> sessions = new ConcurrentHashMap<>();
	private final Map<String, ParticipantState> participants = new ConcurrentHashMap<>();

	public synchronized void register(WebSocketSession session) {
		participants.put(session.getId(), new ParticipantState(null, null));
	}

	public synchronized void remove(WebSocketSession session) {
		ParticipantState state = participants.remove(session.getId());
		if (state == null || state.sessionId() == null) {
			return;
		}

		Set<WebSocketSession> peers = sessions.get(state.sessionId());
		if (peers == null) {
			return;
		}

		peers.remove(session);
		if (peers.isEmpty()) {
			sessions.remove(state.sessionId());
		}
	}

	public synchronized void join(WebSocketSession session, String sessionId, ParticipantRole role) {
		remove(session);

		Set<WebSocketSession> peers = sessions.computeIfAbsent(sessionId, ignored -> new HashSet<>());
		peers.add(session);
		participants.put(session.getId(), new ParticipantState(sessionId, role));
	}

	public synchronized void leave(WebSocketSession session) {
		remove(session);
		participants.put(session.getId(), new ParticipantState(null, null));
	}

	public synchronized int broadcastToPeers(WebSocketSession sender, String rawPayload) {
		ParticipantState state = participants.get(sender.getId());
		if (state == null || state.sessionId() == null) {
			return 0;
		}

		Set<WebSocketSession> peers = sessions.get(state.sessionId());
		if (peers == null) {
			return 0;
		}

		int delivered = 0;
		TextMessage payload = new TextMessage(rawPayload);

		for (WebSocketSession peer : peers) {
			if (peer.getId().equals(sender.getId()) || !peer.isOpen()) {
				continue;
			}

			try {
				peer.sendMessage(payload);
				delivered += 1;
			} catch (IOException ignored) {
				// Transport cleanup is handled by the session close/error callbacks.
			}
		}

		return delivered;
	}
}